proofs/alf/cvc5/programs/Strings.smt3

```diff
@@ -587,3 +587,147 @@
+    (($str_arith_entail_simple n1)          (alf.not (alf.is_neg n1)))
+  )
+)
+
+; program: $str_arith_entail_simple_pred
+; args:
+; - F Bool: A relation over integer terms.
+; return: >
+;   true if we can show that F holds using the fact that length
+;   is non-negative and basic properties of arithmetic. We first
+;   normalize the polynomial corresponding to F and check this
+;   using the $str_arith_entail_simple program above.
+(program $str_arith_entail_simple_pred ((n Int) (m Int))
+  (Bool) Int
+  (
+    (($str_arith_entail_simple_pred (>= n m))  ($str_arith_entail_simple ($arith_poly_to_term (- n m))))
+    (($str_arith_entail_simple_pred (> n m))   ($str_arith_entail_simple ($arith_poly_to_term (- n (+ m 1)))))
+  )
+)
+
+; program: $str_arith_entail_is_approx_len
+; args:
+; - s String: The string whose length we are considering.
+; - n Int: An integer that may under (resp. over) approximate the length of s when b is true (resp. false).
+; - b Bool: Whether we are checking that n is an under (resp. over) approximation.
+; return: >
+;   true if we can show that (>= (str.len s) n) (or <= if b is false) holds using approximation
+;   cases as defined by Figure 2 of "High-Level Abstractions for Simplifying Extended String Constraints
+;   in SMT", CAV 2019. When applicable, we also rely on calls that reason that string
+;   length is non-negative and basic properties of arithmetic.
+(program $str_arith_entail_is_approx_len ((s String) (t String) (r String) (n1 Int) (n2 Int) (n Int) (isUnder Bool))
+  (String Int Bool) Bool
+  (
+    (($str_arith_entail_is_approx_len (str.substr s n1 n2) n isUnder)
+        (let ((npm (+ n1 n2)))
+        (let ((lx (str.len s)))
+        (alf.ite (alf.is_eq n n2)
+          (alf.ite isUnder
+            (alf.and ($str_arith_entail_simple n1) ($str_arith_entail_simple_pred (>= lx npm)))
+            ($str_arith_entail_simple n2))
+        (alf.ite (alf.is_eq n (str.len s))
+          (alf.not isUnder)
+        (alf.ite (alf.is_eq n (- (str.len s) n1))
+          (alf.ite isUnder
+            (alf.and ($str_arith_entail_simple n1) ($str_arith_entail_simple_pred (>= npm lx)))
+            ($str_arith_entail_simple_pred (>= lx n1)))
+          false))))))
+    (($str_arith_entail_is_approx_len (str.replace s t r) n isUnder)
+        (let ((ls (str.len s)))
+        (let ((lt (str.len t)))
+        (let ((lr (str.len r)))
+        (alf.ite (alf.is_eq n ls)
+          (alf.ite isUnder
+            (alf.or ($str_arith_entail_simple_pred (>= lr lt)) ($str_arith_entail_simple_pred (>= lr ls)))
+            ($str_arith_entail_simple_pred (>= lt lr)))
+        (alf.ite (alf.is_eq n (+ ls lr))
+          (alf.not isUnder)
+        (alf.ite (alf.is_eq n (- ls lt))
+          isUnder
+          false)))))))
+    (($str_arith_entail_is_approx_len (str.from_int n1) n isUnder)
+        (alf.ite (alf.is_eq n n1)
+          (alf.and (alf.not isUnder) ($str_arith_entail_simple n1))
```

Comment:
Note that we never need to use 0 as an approximation since it is handled directly in the "core" entailment method, which has the knowledge that length is non-negative. 1 is only an under-approximation if the integer is non-negative (else str.from_int is the empty string).

For reference, the code is here: https://github.com/cvc5/cvc5/blob/main/src/theory/strings/arith_entail.cpp#L684